File: src/opentrace_ml/trace.py
"""Privacy-aware preparation of GPS traces for downstream processing."""

from __future__ import annotations

import hashlib
import hmac
import math
from collections.abc import Sequence
from dataclasses import dataclass
from itertools import pairwise

from .geo import haversine_distance_m
from .models import GeoPoint


@dataclass(frozen=True, slots=True)
class TraceCleaningConfig:
    """Deterministic limits applied at the private trace-ingestion boundary."""

    max_speed_m_s: float = 70.0
    min_points: int = 2

    def __post_init__(self) -> None:
        if not math.isfinite(self.max_speed_m_s) or self.max_speed_m_s <= 0:
            raise ValueError("max_speed_m_s must be a positive finite value")
        if self.min_points < 2:
            raise ValueError("min_points must be at least 2")


@dataclass(frozen=True, slots=True)
class TraceCleaningReport:
    """Aggregate counts safe to log without exposing trace coordinates."""

    input_points: int
    output_points: int
    duplicate_points_removed: int
    speed_outliers_removed: int


@dataclass(frozen=True, slots=True)
class PreparedTrace:
    """A consented, pseudonymized trace ready for private downstream processing."""

    trip_id: str
    points: tuple[GeoPoint, ...]
    report: TraceCleaningReport


def pseudonymize_trip_id(trip_id: str, secret_key: str | bytes) -> str:
    """Return a deterministic HMAC pseudonym without retaining the raw trip ID."""

    normalized_id = trip_id.strip()
    if not normalized_id:
        raise ValueError("trip_id cannot be empty")
    key = secret_key.encode("utf-8") if isinstance(secret_key, str) else secret_key
    if len(key) < 16:
        raise ValueError("secret_key must contain at least 16 bytes")
    digest = hmac.new(
        key,
        b"opentrace-trip-v1\x00" + normalized_id.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()
    return f"trip_{digest}"


def _clean_trace(
    points: Sequence[GeoPoint], config: TraceCleaningConfig
) -> tuple[tuple[GeoPoint, ...], TraceCleaningReport]:
    if len(points) < config.min_points:
        raise ValueError(f"Trace must contain at least {config.min_points} points")

    ordered = list(points)
    if any(
        current.timestamp_seconds > following.timestamp_seconds
        for current, following in pairwise(ordered)
    ):
        raise ValueError("GPS points must be ordered by timestamp")

    kept = [ordered[0]]
    duplicates_removed = 0
    speed_outliers_removed = 0

    for point in ordered[1:]:
        previous = kept[-1]
        duration = point.timestamp_seconds - previous.timestamp_seconds
        distance_m = haversine_distance_m(previous, point)

        if duration == 0:
            if distance_m == 0:
                duplicates_removed += 1
                continue
            raise ValueError("Different GPS positions cannot share the same timestamp")

        if distance_m / duration > config.max_speed_m_s:
            speed_outliers_removed += 1
            continue
        kept.append(point)

    if len(kept) < config.min_points:
        raise ValueError("Trace contains too few usable points after cleaning")

    start = kept[0].timestamp_seconds
    normalized = tuple(
        GeoPoint(point.latitude, point.longitude, point.timestamp_seconds - start)
        for point in kept
    )
    report = TraceCleaningReport(
        input_points=len(ordered),
        output_points=len(normalized),
        duplicate_points_removed=duplicates_removed,
        speed_outliers_removed=speed_outliers_removed,
    )
    return normalized, report


def prepare_trace(
    points: Sequence[GeoPoint],
    *,
    trip_id: str,
    secret_key: str | bytes,
    consent_granted: bool,
    config: TraceCleaningConfig | None = None,
) -> PreparedTrace:
    """Enforce consent, pseudonymize the trip ID, and clean a private trace.

    Callers must pass a trip-scoped identifier, never a hardware device ID. The
    returned coordinates remain sensitive and must not be logged or published.
    """

    if consent_granted is not True:
        raise PermissionError("Explicit consent is required before preparing a GPS trace")
    cleaned, report = _clean_trace(points, config or TraceCleaningConfig())
    return PreparedTrace(
        trip_id=pseudonymize_trip_id(trip_id, secret_key),
        points=cleaned,
        report=report,
    )

File: go/trace.go
package opentrace

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"strings"
)

const tripPseudonymDomain = "opentrace-trip-v1\x00"

// TraceCleaningConfig controls deterministic filtering at the private trace
// boundary.
type TraceCleaningConfig struct {
	MaxSpeedMetersPerSecond float64 `json:"max_speed_m_s"`
	MinPoints               int     `json:"min_points"`
}

// DefaultTraceCleaningConfig returns the cross-language defaults.
func DefaultTraceCleaningConfig() TraceCleaningConfig {
	return TraceCleaningConfig{MaxSpeedMetersPerSecond: 70, MinPoints: 2}
}

// Validate checks cleaning thresholds without processing coordinates.
func (config TraceCleaningConfig) Validate() error {
	if !isFinite(config.MaxSpeedMetersPerSecond) || config.MaxSpeedMetersPerSecond <= 0 {
		return errors.New("opentrace: max speed must be positive and finite")
	}
	if config.MinPoints < 2 {
		return errors.New("opentrace: minimum points must be at least 2")
	}
	return nil
}

// TraceCleaningReport contains aggregate counts safe to log.
type TraceCleaningReport struct {
	InputPoints            int `json:"input_points"`
	OutputPoints           int `json:"output_points"`
	DuplicatePointsRemoved int `json:"duplicate_points_removed"`
	SpeedOutliersRemoved   int `json:"speed_outliers_removed"`
}

// PreparedTrace is a consented and pseudonymized private trace. Coordinates
// are intentionally unexported; Points returns a defensive copy.
type PreparedTrace struct {
	tripID string
	points []GeoPoint
	report TraceCleaningReport
}

// TripID returns the trip-scoped HMAC pseudonym.
func (trace PreparedTrace) TripID() string { return trace.tripID }

// Points returns a copy of sensitive intermediate coordinates. Callers must not
// log or publish the returned slice.
func (trace PreparedTrace) Points() []GeoPoint {
	return append([]GeoPoint(nil), trace.points...)
}

// Len returns the number of retained points without copying coordinates.
func (trace PreparedTrace) Len() int { return len(trace.points) }

// Report returns aggregate trace-cleaning counts.
func (trace PreparedTrace) Report() TraceCleaningReport { return trace.report }

// String redacts coordinates from ordinary diagnostic formatting. Pseudonyms
// and coordinates remain sensitive even after explicit preparation.
func (trace PreparedTrace) String() string {
	return fmt.Sprintf("PreparedTrace{points:%d, coordinates:redacted}", trace.Len())
}

// GoString also redacts coordinates from fmt's %#v formatting.
func (trace PreparedTrace) GoString() string { return trace.String() }

// PseudonymizeTripID produces the exact HMAC-SHA256 pseudonym used by the
// Python package. Runtime is O(len(tripID)).
func PseudonymizeTripID(tripID string, secretKey []byte) (string, error) {
	normalized := strings.TrimSpace(tripID)
	if normalized == "" {
		return "", errors.New("opentrace: trip ID cannot be empty")
	}
	if len(secretKey) < 16 {
		return "", errors.New("opentrace: secret key must contain at least 16 bytes")
	}
	mac := hmac.New(sha256.New, secretKey)
	_, _ = mac.Write([]byte(tripPseudonymDomain))
	_, _ = mac.Write([]byte(normalized))
	return "trip_" + hex.EncodeToString(mac.Sum(nil)), nil
}

// TracePreparationOptions makes consent and the pseudonym key explicit at the
// call boundary.
type TracePreparationOptions struct {
	TripID         string
	SecretKey      []byte
	ConsentGranted bool
	Config         TraceCleaningConfig
}

// PrepareTrace enforces consent, rejects ambiguous timestamps, removes exact
// duplicates and implausible speed jumps, and normalizes timestamps in O(n).
func PrepareTrace(points []GeoPoint, options TracePreparationOptions) (PreparedTrace, error) {
	if !options.ConsentGranted {
		return PreparedTrace{}, ErrConsentRequired
	}
	config := options.Config
	if config == (TraceCleaningConfig{}) {
		config = DefaultTraceCleaningConfig()
	}
	if err := config.Validate(); err != nil {
		return PreparedTrace{}, err
	}
	if len(points) < config.MinPoints {
		return PreparedTrace{}, fmt.Errorf("opentrace: trace must contain at least %d points", config.MinPoints)
	}
	if err := validateSourceTrace(points); err != nil {
		return PreparedTrace{}, err
	}

	kept := make([]GeoPoint, 0, len(points))
	kept = append(kept, points[0])
	duplicatesRemoved, speedOutliersRemoved := 0, 0
	for _, point := range points[1:] {
		previous := kept[len(kept)-1]
		duration := point.TimestampSeconds - previous.TimestampSeconds
		distance := haversineUnchecked(previous, point)
		if duration == 0 {
			duplicatesRemoved++
			continue
		}
		if distance/duration > config.MaxSpeedMetersPerSecond {
			speedOutliersRemoved++
			continue
		}
		kept = append(kept, point)
	}
	if len(kept) < config.MinPoints {
		return PreparedTrace{}, errors.New("opentrace: trace contains too few usable points after cleaning")
	}

	start := kept[0].TimestampSeconds
	for i := range kept {
		kept[i].TimestampSeconds -= start
	}
	pseudonym, err := PseudonymizeTripID(options.TripID, options.SecretKey)
	if err != nil {
		return PreparedTrace{}, err
	}
	return PreparedTrace{
		tripID: pseudonym,
		points: kept,
		report: TraceCleaningReport{
			InputPoints:            len(points),
			OutputPoints:           len(kept),
			DuplicatePointsRemoved: duplicatesRemoved,
			SpeedOutliersRemoved:   speedOutliersRemoved,
		},
	}, nil
}

func validateSourceTrace(points []GeoPoint) error {
	for i, point := range points {
		if err := point.Validate(); err != nil {
			return fmt.Errorf("opentrace: point %d: %w", i, err)
		}
		if i == 0 {
			continue
		}
		previous := points[i-1]
		if previous.TimestampSeconds > point.TimestampSeconds {
			return errors.New("opentrace: GPS points must be ordered by timestamp")
		}
		if previous.TimestampSeconds == point.TimestampSeconds &&
			(previous.Latitude != point.Latitude || previous.Longitude != point.Longitude) {
			return errors.New("opentrace: different GPS positions cannot share the same timestamp")
		}
	}
	return nil
}

func validTripPseudonym(value string) bool {
	if len(value) != len("trip_")+sha256.Size*2 || !strings.HasPrefix(value, "trip_") {
		return false
	}
	decoded, err := hex.DecodeString(strings.TrimPrefix(value, "trip_"))
	return err == nil && len(decoded) == sha256.Size && value == strings.ToLower(value)
}
